packages/babel-cli/src/babel/file.ts

```diff
@@ -19,66 +18,66 @@ export default async function ({
   babelOptions,
 }: CmdOptions): Promise<void> {
   function buildResult(fileResults: Array<any>): CompilationOutput {
-    const map = new sourceMap.SourceMapGenerator({
-      file:
-        cliOptions.sourceMapTarget ||
-        path.basename(cliOptions.outFile || "") ||
-        "stdout",
-      sourceRoot: babelOptions.sourceRoot,
-    });
+    const mapSections = [];
 
     let code = "";
     let offset = 0;
 
     for (const result of fileResults) {
       if (!result) continue;
 
-      code += result.code + "\n";
-
-      if (result.map) {
-        const consumer = new TraceMap(result.map);
-
-        eachMapping(consumer, mapping => {
-          map.addMapping({
-            generated: {
-              line: mapping.generatedLine + offset,
-              column: mapping.generatedColumn,
-            },
-            source: mapping.source,
-            original:
-              mapping.source == null
-                ? null
-                : {
-                    line: mapping.originalLine,
-                    column: mapping.originalColumn,
-                  },
-          });
-        });
-
-        const { resolvedSources, sourcesContent } = consumer;
-        sourcesContent?.forEach((content, i) => {
-          if (content === null) return;
-          map.setSourceContent(resolvedSources[i], content);
-        });
+      mapSections.push({
+        offset: { line: offset, column: 0 },
+        map: result.map || emptyMap(),
+      });
 
-        offset = code.split("\n").length - 1;
-      }
+      code += result.code + "\n";
+      offset += countNewlines(result.code) + 1;
     }
 
+    const map = new AnyMap({
+      version: 3,
+      file:
+        cliOptions.sourceMapTarget ||
+        path.basename(cliOptions.outFile || "") ||
+        "stdout",
+      sections: mapSections,
+    });
+    // For some reason, the spec doesn't allow sourceRoot when constructing a
+    // sectioned sorucemap. But AllMap returns a regular sourcemap, we can
+    // freely add to with a sourceRoot.
+    map.sourceRoot = babelOptions.sourceRoot;
+
     // add the inline sourcemap comment if we've either explicitly asked for inline source
     // maps, or we've requested them without any output file
     if (
       babelOptions.sourceMaps === "inline" ||
       (!cliOptions.outFile && babelOptions.sourceMaps)
     ) {
-      code += "\n" + convertSourceMap.fromObject(map).toComment();
+      code += "\n" + convertSourceMap.fromObject(encodedMap(map)).toComment();
     }
 
     return {
       map: map,
       code: code,
     };
   }
+  function countNewlines(code: string): number {
+    let count = 0;
+    let index = -1;
+    while ((index = code.indexOf("\n", index + 1)) !== -1) {
```

Comment:
I think we need to handle all the possible newlines. Even if we always print `\n`, multiline template literals may still contain different line terminators.

Reply:
This is just a faster version of the `code.split("\n").length ` that was there before. We can add more newline support, could we wait until a followup?